File: src/day2.rs
use crate::{Error, get_input_lines};

const INPUT: &str = "problems/problem2";

#[derive(Debug, Clone, Copy)]
enum Instruction {
    Forward(i64),
    Down(i64),
    Up(i64)
}

impl TryFrom<&str> for Instruction {
    type Error = crate::Error;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        let mut split = value.split_ascii_whitespace();
        let first = split.next().ok_or(Error::InvalidInput)?;
        let second = split.next().ok_or(Error::InvalidInput)?;

        let nr = second.parse()?;

        match first {
            "forward" => Ok(Instruction::Forward(nr)),
            "down" => Ok(Instruction::Down(nr)),
            "up" => Ok(Instruction::Up(nr)),
            _ => Err(Error::InvalidInput),
        }
    }
}

pub fn solve() -> Result<(), crate::Error> {
    let instructions = get_input_lines(INPUT)?
        .filter_map(|s| Instruction::try_from(s.as_str()).ok())
        .collect::<Vec<Instruction>>();

    // Part 1
    let mut pos = 0;
    let mut depth = 0;

    for &inst in &instructions {
        match inst {
            Instruction::Down(x) => depth += x,
            Instruction::Up(x) => depth -= x,
            Instruction::Forward(x) => pos += x
        }
    }

    println!("Problem 1: {}", pos * depth);

    // Part 2
    let mut pos = 0;
    let mut depth = 0;
    let mut aim = 0;

    for inst in instructions {
        match inst {
            Instruction::Down(x) => aim += x,
            Instruction::Up(x) => aim -= x,
            Instruction::Forward(x) => {
                pos += x;
                depth += aim * x;
            }
        }
    }

    println!("Problem 2: {}", pos * depth);

    Ok(())
}

File: python/day2.py
def solve():
    with open("problems/problem2") as f:
        lines = f.readlines()

    pos1 = 0
    depth1 = 0

    aim2 = 0
    pos2 = 0
    depth2 = 0

    for line in lines:
        instruction, x = line.split()
        x = int(x)

        if instruction == 'forward':
            pos1 += x

            pos2 += x
            depth2 += aim2 * x
        elif instruction == 'down':
            depth1 += x

            aim2 += x
        elif instruction == 'up':
            depth1 -= x

            aim2 -= x

    print("Problem 1:", pos1 * depth1)
    print("Problem 2:", pos2 * depth2)

if __name__ == '__main__':
    solve()